# Đồ Án Lập Trình cho Khoa Học Dữ Liệu

## Thành viên
| Mã số sinh viên | Học và tên                                           |
|:---------------:|:-----------------------------------------------------|
|18120225         | [Trần Thanh Phúc](https://github.com/PhucThanh)      |
|18120221         | [Nguyễn Điền Thanh Phong](https://github.com/sgrayk) |

---

## Môi trường code

In [36]:
!which python 

/home/workspace/miniconda3/envs/p4ds-env/bin/python


---

## Các thư viện sử dụng

In [37]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

---

## Thu thập dữ liệu

TV Shows and Movies listed on Netflix

Dataset này bao gồm tv shows và movies khả dụng trên Netflix của năm 2019. Tập dữ liệu được thu thập từ Flixable, một công cụ tìm kiếm Netflix của bên thứ ba.

Vào năm 2018, họ đã phát hành một báo cáo thú vị cho thấy số lượng chương trình truyền hình trên Netflix đã tăng gần gấp ba lần kể từ năm 2010. Số lượng phim của dịch vụ phát trực tuyến đã giảm hơn 2.000 tựa kể từ năm 2010, trong khi số lượng chương trình truyền hình của nó tăng gần gấp ba lần. Sẽ rất thú vị khi khám phá xem tất cả những thông tin chi tiết khác có thể thu được từ cùng một tập dữ liệu.

Some of the interesting questions (tasks) which can be performed on this dataset -

    Understanding what content is available in different countries
    Identifying similar content by matching text-based features
    Network analysis of Actors / Directors and find interesting insights
    Is Netflix has increasingly focusing on TV rather than movies in recent years.


---

## Khám phá dữ liệu

### Đọc dữ liệu từ file csv vào dataframe

In [38]:
netflix_df = pd.read_csv('./datasheet/netflix_titles.csv')
netflix_df.head()

show_id     type  title           director  \
0      s1  TV Show     3%                NaN   
1      s2    Movie   7:19  Jorge Michel Grau   
2      s3    Movie  23:59       Gilbert Chan   
3      s4    Movie      9        Shane Acker   
4      s5    Movie     21     Robert Luketic   

                                                cast        country  \
0  João Miguel, Bianca Comparato, Michel Gomes, R...         Brazil   
1  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
2  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
3  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
4  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   

          date_added  release_year rating   duration  \
0    August 14, 2020          2020  TV-MA  4 Seasons   
1  December 23, 2016          2016  TV-MA     93 min   
2  December 20, 2018          2011      R     78 min   
3  November 16, 2017          2009  PG-13     80 min   
4    January 1, 2020          2008  PG-13    123 min   

                                           listed_in  \
0  International TV Shows, TV Dramas, TV Sci-Fi &...   
1                       Dramas, International Movies   
2                Horror Movies, International Movies   
3  Action & Adventure, Independent Movies, Sci-Fi...   
4                                             Dramas   

                                         description  
0  In a future where the elite inhabit an island ...  
1  After a devastating earthquake hits Mexico Cit...  
2  When an army recruit is found dead, his fellow...  
3  In a postapocalyptic world, rag-doll robots hi...  
4  A brilliant group of students become card-coun...

### Số dữ liệu và số cột trong Dữ liệu

In [39]:
num_rows = netflix_df.shape[0]
num_cols = netflix_df.shape[1]

print("Số dữ liệu trong DataFrame: ", num_rows)
print("Số cột dữ liệu trong DataFrame: ", num_cols)

Số dữ liệu trong DataFrame:  7787
Số cột dữ liệu trong DataFrame:  12


### Kiểm tra sự trùng lặp trong dữ liệu

In [40]:
have_duplicated_rows = np.sum(netflix_df.duplicated(keep=False)) != 0

print("Dữ liệu có bị trùng lặp: ", have_duplicated_rows)

Dữ liệu có bị trùng lặp:  False


### Ý nghĩa mỗi cột trong Dữ liệu

- **show_id**: các giá trị đơn lập giúp xác định một Chương trình. - str
- **type**: loại chương trình của Netflix, là Movie hoặc TV Show.  - str
- **title**: tên chương trình được phát trên Netflix. - str
- **director**: đạo diễn chương trình - str
- **cast**: danh sách các diễn viên trong Chương trình, ngăn cách bởi dấu "," - str
- **country**: tên nước sản xuất - str
- **date_added**: thời gian được thêm vào Netflix - datetime
- **release_year**: thời gian sản xuất chương trình - int
- **rating**: đánh giá phim thuộc thể loại nào và mức độ của nó - str
- **duration**: thời lượng chương trình, có thể kéo dài trong hơn một mùa,... - str
- **listed_in**: thể loại chương trình - str
- **description**: mô tả chương trình - str

### Kiểm tra loại dữ liệu của mỗi cột có đơn nhất không?

In [41]:
col_dtypes = netflix_df.dtypes

col_dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

**Kiểm tra kiểu thật sự của những cột có dtype là Object**

`netflix_df` có 11 cột có `dtype` là `object`: "show_id", "type", "title", "director", "cast", "country", "date_added", "rating", "duration", "listed_in", "description". Ta sẽ kiểm tra nội bộ mỗi cột này thực sự là các kiểu dữ liệu nào.

In [42]:
def open_object_dtype(s):
    
    dtypes = set(s.apply(type))
    
    return dtypes

for col in netflix_df.columns:
    print("Column %12s có kiểu dữ liệu: %s" % (col, open_object_dtype(netflix_df[col])))

Column      show_id có kiểu dữ liệu: {<class 'str'>}
Column         type có kiểu dữ liệu: {<class 'str'>}
Column        title có kiểu dữ liệu: {<class 'str'>}
Column     director có kiểu dữ liệu: {<class 'str'>, <class 'float'>}
Column         cast có kiểu dữ liệu: {<class 'str'>, <class 'float'>}
Column      country có kiểu dữ liệu: {<class 'str'>, <class 'float'>}
Column   date_added có kiểu dữ liệu: {<class 'str'>, <class 'float'>}
Column release_year có kiểu dữ liệu: {<class 'int'>}
Column       rating có kiểu dữ liệu: {<class 'str'>, <class 'float'>}
Column     duration có kiểu dữ liệu: {<class 'str'>}
Column    listed_in có kiểu dữ liệu: {<class 'str'>}
Column  description có kiểu dữ liệu: {<class 'str'>}


**Vấn đề cần tiền xử lý:**
Theo thông tin ta nhận được ở trên, ta có những đánh giá sau: 
- Cột "date_added" đang có dtype là `str` và `float`. Để đúng về ý nghĩa, ta sẽ thực hiện bước tiền xử lý là chuyển sang dạng `datetime`.
- Cột "director", "cast", "country" và "rating" có dtype gồm 2 loại là `str` và `float`, ta sẽ tiền xử lý để thống nhất `dtype` là `str` 
- Hầu hết các cột có giá trị `float` là vì các cột có giá trị `nan`

---

## Tiền xử lý

### Xử lý cột `date_added`

**Kiểm tra tỷ lệ mất mát dữ liệu:**
- Do cột có các cột có giá trị nan nên ta sẽ xem xét tỷ lệ này có ảnh hưởng gì đễn dữ liệu không

In [43]:
def fMissingRatio(name_col):
    print("Tỷ lệ mất mát dữ liệu của cột %s là %3.3f." % (name_col, round(netflix_df[name_col].isna().mean() * 100,3)))
    
fMissingRatio('date_added')

Tỷ lệ mất mát dữ liệu của cột date_added là 0.128.


**Giải quyết:**
- Do cột có các cột có giá trị nan và các dòng dữ liệu không liên quan tới nhau cộng thêm tỷ lệ mất mát nhỏ nên ta sẽ loại bỏ chúng và dữ liệu không bị ảnh hưởng.

In [44]:
netflix_df.dropna(subset = ['date_added'], inplace=True)
print("Số dòng dữ liệu: ", netflix_df.shape)
fMissingRatio('date_added')

Số dòng dữ liệu:  (7777, 12)
Tỷ lệ mất mát dữ liệu của cột date_added là 0.000.


**Kiểm tra:**
- Vậy DataFrame trên cột `date_added` đã sạch.

**Chuyển dtype của cột "date_added" sang datetime**

In [45]:
netflix_df['date_added'] = pd.to_datetime(netflix_df['date_added'].apply(str.strip), format='%B %d, %Y')

open_object_dtype(netflix_df['date_added'])

{pandas._libs.tslibs.timestamps.Timestamp}

### Xử lý các cột `director`, `cast`, `country` và `rating`

**Chuyển dtype của cột "director", "cast", "country" và "rating" sang str**

Vì `dtype` các cột trên đều là `str` và `nan` nên khi chuyển sang `str`, ta sẽ thay thế các giá trị `nan` trong các cột với chuỗi rỗng.

In [46]:
netflix_df['director'] = netflix_df['director'].replace(to_replace=np.nan, value="")
netflix_df['cast'] = netflix_df['cast'].replace(to_replace=np.nan, value="")
netflix_df['country'] = netflix_df['country'].replace(to_replace=np.nan, value="")
netflix_df['rating'] = netflix_df['rating'].replace(to_replace=np.nan, value="")

### Kiểm tra tính đồng nhất kiểu dữ liệu trong DataFrame 

### Kiểm tra tỷ lệ mất mát dữ liệu trong DataFrame 

In [47]:
for col in netflix_df.columns:
    fMissingRatio(col)

Tỷ lệ mất mát dữ liệu của cột show_id là 0.000.
Tỷ lệ mất mát dữ liệu của cột type là 0.000.
Tỷ lệ mất mát dữ liệu của cột title là 0.000.
Tỷ lệ mất mát dữ liệu của cột director là 0.000.
Tỷ lệ mất mát dữ liệu của cột cast là 0.000.
Tỷ lệ mất mát dữ liệu của cột country là 0.000.
Tỷ lệ mất mát dữ liệu của cột date_added là 0.000.
Tỷ lệ mất mát dữ liệu của cột release_year là 0.000.
Tỷ lệ mất mát dữ liệu của cột rating là 0.000.
Tỷ lệ mất mát dữ liệu của cột duration là 0.000.
Tỷ lệ mất mát dữ liệu của cột listed_in là 0.000.
Tỷ lệ mất mát dữ liệu của cột description là 0.000.


Như vậy, dữ liệu đã gần như "sạch". 

---

## Khám phá dữ liệu (tiếp tục)

### Với mỗi cột có kiểu dữ liệu dạng numeric, ta cần xem xét sự phân bố các giá trị

Ta sẽ xếp các cột `date_added` vào nhóm numeric, do đó hiện có 2 cột thuộc nhóm numeric: "release_year", "date_added".

Với mỗi cột numeric ta sẽ xem những thông tin mô tả dữ liệu.   

In [48]:
nume_col_profiles_df = netflix_df.loc[:,['release_year', 'date_added']] \
                                 .agg([lambda x: pd.Series.describe(x, datetime_is_numeric=True)])

nume_col_profiles_df.columns = ['release_year', 'date_added']

nume_col_profiles_df

release_year                     date_added
count   7777.000000                           7777
mean    2013.935965  2019-01-02 19:20:57.708627968
std        8.760694                            NaN
min     1925.000000            2008-01-01 00:00:00
25%     2013.000000            2018-02-01 00:00:00
50%     2017.000000            2019-03-08 00:00:00
75%     2018.000000            2020-01-20 00:00:00
max     2021.000000            2021-01-16 00:00:00

Dựa vào bảng giá trị, ta có những nhận xét sơ như sau:
- Nextflix đã thêm những chương trình được sản xuất vào năm 1925 dù bộ phim đầu tiên năm 1895 và Netflix được ra đời năm 1997, vậy thì tại sao Netflix lại chọn bộ phim cũ như vậy?  
- Dữ liệu cuối cùng là năm 2021, có thể người thu thập chỉ cập nhật chọn thời điểm 2008 - 2021 nên dữ liệu có vẻ không đầy đủ. 
- Mức 25% là năm 2013, trong khi 0% là 1925, vậy những bộ phim hay chương trình vào thời điểm trước 2013 khá hạn chế, có thể sau năm 2013 thì Netflix phát triển mạnh lên.
- Trên mỗi mức, ta thấy thời gian khá chênh lệch, có thể Netflix chờ những bộ phim dài tập hoặc chương hoàn tất thì Netflix mới thêm bộ phim hay chương trình vào Netflix. Còn mức 100%, cả 2 cùng năm 2021, có thể chương trình mới nhất là một TV Show.

### Với mỗi cột có kiểu dữ liệu dạng categorical, ta cần xem xét sự phân bố các giá trị

`netflix_df` có 10 cột categorical là "show_id", "type", "title", "director", "cast", "country", "rating", "duration", "listed_in" và "description".

Với mỗi cột categorical, ta sẽ xem xét số lượng giá trị khác nhau, list/array các giá trị khác nhau.

In [76]:
# cate_col_profiles_df = netflix_df.loc[:,['show_id','type','title','director','cast','country','rating','duration','listed_in','description']] \
#                                  .agg([pd.Series.nunique, pd.Series.unique])

# cate_col_profiles_df.index = ['num_diff_vals','diff_vals']

# cate_col_profiles_df

cate_col_profiles_df = netflix_df.loc[:,'cast'] \
                                 .agg([pd.Series.nunique, pd.Series.unique])

cate_col_profiles_df

nunique                                                 6822
unique     [João Miguel, Bianca Comparato, Michel Gomes, ...
Name: cast, dtype: object

Dựa vào bảng giá trị, ta có những nhận xét sơ như sau:
- `show_id` có đủ 7777 giá trị, điều này có thể thấy bình thường.
- `type` gồm 2 loại là TV Show và Movie, đúng như ta đã nói ban đầu, không có gì kỳ lạ.
- `title` gồm 7777 tên đơn lập, đúng như mô tả của người lập ra Dataset này. 
- `description` chỉ có 7759 giá trị thay vì 7777, khá kì lạ khi mô tả phải khác nhau khi tên chương trình và phim khác nhau. 
- `duration` có 216 giá trị, nhưng có thể trích rút ra đơn vị tính từ cột này.
- Có 15 mức độ đánh giá của bảng dữ liệu này được rút ra từ cột `rating`. 
- Từ cột `country`, ta thấy 682 nước có chương trình và bộ phim trong data của Netflix.
- Cột `cast` cho thấy có 955 trường hợp là cùng đóng chung với nhau ở chương trình hoặc bộ phim nào đó, khá thú vị. 
- Những cột khác không có gì đặc biệt.

In [50]:
for x in netflix_df['description']:
    if (len(x) < 10):
        print(x)

---